src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Target workspace addressed by the permanent ID",
+    )
+    required.add_argument(
+        f"--{loader_common_names.KBASE_COLLECTION_ARG_NAME}",
+        type=str,
+        help="Create a collection and link in data to that collection from the overall workspace source data dir",
+    )
+    required.add_argument(
+        f"--{loader_common_names.SOURCE_VER_ARG_NAME}",
+        type=str,
+        help="Create a source version and link in data to that collection from the overall workspace source data dir",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--token_filepath",
+        type=str,
+        help="A file path that stores KBase token",
+    )
+    optional.add_argument(
+        "--env",
+        type=str,
+        choices=loader_common_names.KB_ENV,
+        default="PROD",
+        help="KBase environment, defaulting to PROD",
+    )
+    optional.add_argument(
+        "--upload_file_ext",
+        type=str,
+        nargs="+",
+        default=UPLOAD_FILE_EXT,
+        help="Upload only files that match given extensions",
+    )
+    optional.add_argument(
+        "--keep_job_dir",
+        action="store_true",
+        help="Keep SDK job directory after upload task is completed",
+    )
+    return parser
+
+
+def _get_yaml_file_path(root_dir: str, assembly_dir_name: str) -> str:
+    """
+    Get the uploaded.yaml file path from NCBI/sourcedata directory.
+    """
+    file_path = os.path.join(get_work_dir(root_dir), assembly_dir_name, UPLOADED_YAML)
+    Path(file_path).touch(exist_ok=True)
+    return file_path
+
+
+def _sanitize_data_dir(job_dir: str) -> str:
+    """
+    Create a temporary directory for storing uploaded files.
+    """
+    data_dir = os.path.join(job_dir, "workdir/tmp", DATA_DIR)
+    if os.path.exists(data_dir):
+        shutil.rmtree(data_dir)
+    os.makedirs(data_dir)
+    return data_dir
+
+
+def _get_source_file(assembly_dir: str, assembly_file: str) -> str:
+    """
+    Get the sourcedata file path from the assembly directory.
+    """
+    if not os.path.islink(assembly_dir):
+        raise ValueError(f"{assembly_dir} is not a symlink")
+    src_file = os.path.join(os.readlink(assembly_dir), assembly_file)
+    return src_file
+
+
+def _upload_assembly_to_workspace(
+    conf: Conf,
+    workspace_name: str,
+    file_path: str,
+    assembly_name: str,
+) -> None:
+    """Upload an assembly file to workspace."""
+    success, attempts, max_attempts = False, 0, 3
+    while attempts < max_attempts and not success:
+        try:
+            time.sleep(attempts)
+            conf.asu.save_assembly_from_fasta(
```

Comment:
* use `save_assembly_from_fasta2`, at least, if you're not going to use the bulk method
* Why not get the UPA here?

Reply:
During the performance testing part of this you may want to see if the bulk method improves things

Reply:
> I get upa info after entire upload process is complete. I am curious what benefits I would gain if I get UPA info

It's just that it's returned from the function call so why not just keep that in a mapping or something rather than having to make another WS call later

Reply:
Updated!

Reply:
:+1: